enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1179,14 +1190,56 @@ func (dc *writeCloser) Write(p []byte) (int, error) {
 	return n, nil
 }
 
-// TODO(Maggie): Compress all data above a given size before writing it
+// zstdCompressor compresses bytes before writing them to the nested writer
+type zstdCompressor struct {
+	*ioutil.CustomCommitWriteCloser
+	compressBuf []byte
+	bufferPool  *bytebufferpool.Pool
+}
+
+func (z *zstdCompressor) Write(decompressedBytes []byte) (int, error) {
+	z.compressBuf = compression.CompressZstd(z.compressBuf, decompressedBytes)
+	compressedBytesWritten, err := z.CustomCommitWriteCloser.Write(z.compressBuf)
+	if err != nil {
+		return 0, err
+	}
+
+	metrics.CompressionRatio.With(prometheus.Labels{metrics.CompressionType: "zstd"}).Observe(float64(compressedBytesWritten) / float64(len(decompressedBytes)))
+	if compressedBytesWritten > len(decompressedBytes) {
+		metrics.BadCompressionBufferSize.With(prometheus.Labels{metrics.CompressionType: "zstd"}).Observe(float64(len(decompressedBytes)))
```

Comment:
A large "bad" buffer that is only bad by 1 byte, and a medium-sized buffer that is bad by 5000% can potentially both show up in the same bucket.

Would it be more useful to record `(compressedSize - decompressedSize)` here?

Reply:
(Or are you trying to measure whether it's true that it's mostly the smaller buffers that have bad compression? If that's the case, fine to leave as-is)

Reply:
Yeah was trying to validate that smaller buffers have bad compression, so we can tune at what point we want to compress the data. I think the other compression ratio metric will help us track "badness", so we don't need to record (compressedSize - decompressedSize)

Also made the size at which we start compressing a flag, so it's more configurable if we notice that it's too small